### DISCUSSION TITLE
Symfony 6 Doctrine without Injection Dependency

### DISCUSSION THREAD

**Discussion by syber-omnichannel** [symfony/symfony#45357] [Q&A]
Hi, I have a basic question that I can't find a solution. Considering that I use Symfony 6, how would I be able to access the $doctrine variable in the example below without doing dependency injection, without __constructor()?

```
class AccountController extends AbstractController
{
   public function onUpdate()
   {
      $orderEntity = $doctrine->getRepository(Order::class)->findOneBy(['id' => 1]);
   }
}
```

**Comment by ahundiak** [comment]
I guess it would depend in part on why you want to do such a thing with those restrictions.  You could access it globally:
```
public function onUpdate() {
        $kernel = $GLOBALS['app'];
        $container = $kernel->getContainer();
        $doctrine = $container->get('doctrine');
```
But as you might imagine, using globals is frowned upon.

If you are trying to compensate for the fact that the doctrine service was dropped from the default list of the AbstractController's subscribed services then you could add it back in by adding AccountController::getSubscribedServices though technically that does use DI.

You could also make yourself a DoctrineTrait and use the depends annotation to cleanly inject the service.  But again DI.

I suppose you could even do a Laravel Facade sort of thing and make yourself a Doctrine singleton of sorts.

It all depends on why you feel the need to do such a thing.

**Reply by syber-omnichannel** [reply]
I believe I have a concept error because I'm starting now, because I didn't want to use DI so I wouldn't have to pass it as a parameter if I were going to create a new instance of AccountController in another controller, that is, if I add in the AccountController constructor the ManagerRegistry DI (`__construct(private ManagerRegistry $doctrine)`), whenever I instantiate a new AccountController, I will have to pass a ManagerRegistry like this

`$account = new AccountController($managerRegistry)`

Do you understand my question and what I want to avoid? I want to call like this:

`$account = new AccountController()`

And inside AccountController, when necessary, get a ManagerRegistry intance, you know?

**Reply by ahundiak** [reply]
Yep.  Definitely not understanding some basic Symfony concepts. 

First off, controllers are only instantiated by the framework as it processes the request.  You would never instantiate one yourself.  In fact, except for data objects, you pretty much never instantiate an object using the new operator when using the Symfony framework.  

Secondly, I am assuming you have some Accounted oriented functionality (such as onUpdate) that you want to run in multiple places.  If so you place this functionality into it's own service and then inject it as needed.

It is perhaps a different way of thinking about how to write code.  All I can really suggest is reading the docs on the service container and following some of the examples.  It's pretty straight forward once you get the hang of it.

I took a peek at your reply to chyshyya.  The getDoctrine method has been removed in 6.0 so if you want to use the Doctrine services then you do need to inject them.  However, `$x = new AccountController(); $x->onUpdate();` absolutely did not work in previous versions.

**Reply by syber-omnichannel** [reply]
Thanks! That's what I imagined then, I had the wrong concepts... I'll take a look at the documentation you suggested.

**Comment by chyshyya** [comment]
```
public function onUpdate(EntityManagerInterface $entityManager)
{
    $orderEntity = $entityManager->getRepository(Order::class)->findOneBy(['id' => 1]);
}
```
but the right way is to autowire Repository itself:
```
public function onUpdate(OrderRepository $orderRepository)
{
    $orderEntity = $orderRepository->find(1);
}
```

**Reply by syber-omnichannel** [reply]
Thank you very much for the tip about the Repository, but I still have difficulty, because in the case that you suggested, if I call this method in any part of the system, I will have to pass an OrderRepository, see an example of what I imagine (and I don't know if the concept is correct!)

```
// controller/AccountController.php
class AccountController extends AbstractController
{
   public function onUpdate(OrderRepository $orderRepository)
   {
      $orderEntity = $orderRepositoty->find(1);
   }
}
```

Now imagine I have another controller

```
// controller/OrderController.php
class OrderController extends AbstractController
{
   public function create()
   {
      $orderId = 1;
      ....
      $account = new AccountController();
      $account->onUpdate($orderId);
      ...
   }
}
```

Note that for the OrderController controller it doesn't need to know what onUpdate() is, it just wants to tell the onUpdate method that it should update an account linked to the orderId. 

See that in OrderController OrderRepository is not used, but there would have to be an instance of it to pass in the onUpdate() parameter, you know?

In previous versions of Symfony I could do the following...

```
   public function onUpdate()
   {
      $orderEntity = $this->getDoctrine()->getRepository(Order::class)->findOneBy(['id' => 1]);
   }
```

That is, I could call onUpdate anywhere without having to pass any parameters.

Ignore the logic of the names, it's just an example to try to explain where my question is.
